### DISCUSSION TITLE
WindowsAppSDK (WinUI3) shutdown application before mainwindow is shown?

### DISCUSSION THREAD

**Discussion by nathanAjacobs** [microsoft/WindowsAppSDK#3278] [Q&A] [answered]
In WindowsAppSDK (WinUI 3) is it possible to shutdown the application specifically in the OnLaunched override before the main window is shown? Application.Exit() does not work until the main window is shown.

**Comment by ghost1372** [comment]
Mybe killing process

**Comment by DarranRowe** [comment, marked as answer]
The documentation for [Microsoft.UI.Xaml.Application.Start](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.xaml.application.start?view=windows-app-sdk-1.2#microsoft-ui-xaml-application-start(microsoft-ui-xaml-applicationinitializationcallback)) gives a bit of a hint. It states:

"This runs a message pump internally, and does not return until the application shuts down."

What it means by this "message pump" is the GetMessage/DispatchMessage loop. There is more evidence of this in the dispatcher queue documentation where [DispatcherQueueController.CreateOnCurrentThread](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.dispatching.dispatcherqueuecontroller.createoncurrentthread?view=windows-app-sdk-1.2#microsoft-ui-dispatching-dispatcherqueuecontroller-createoncurrentthread) explicitly states that it uses a USER32 message loop. The dispatcher queue is what does the WinUI 3 event processing. This means that all you have to do is exit this loop, and there is a very simple but stupid way of doing that. [PostQuitMessage](https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-postquitmessage). This will cause GetMessage to exit the loop.

Of course, if you are working in C# then you will have to platform invoke, but that should be trivial.

**Reply by nathanAjacobs** [reply]
Thanks, good to know!

**Reply by nathanAjacobs** [reply]
It would be nice if there was a static Shutdown method in the Application class that essentially does this, but this will do. Thanks again.
